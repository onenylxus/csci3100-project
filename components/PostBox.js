// Import
import React from 'react';
import { Alert, Text, View, TouchableOpacity, Image } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { Col, Grid } from 'react-native-easy-grid';
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome';
import {
  faCommentAlt,
  faEdit,
  faExclamation,
  faTrashAlt,
} from '@fortawesome/free-solid-svg-icons';
import CommentContainer from './CommentContainer';
import LikeAndDislike from './LikeAndDislike';
import Source from '../assets/source';
import Style from '../assets/style';

// Export Post Box
export default function PostBox({ post, showButton }) {
  const navigation = useNavigation();

  const [showComment, setShowComment] = React.useState(false);

  const status = React.useRef(0);
  const postUsername = React.useRef(post.username);
  const date = React.useRef(new Date(post.timestamp));
  const monthString = React.useRef(date.current.getMonth() + 1);
  const dateString = React.useRef(
    date.current.getHours().toString().padStart(2, '0') +
      ':' +
      date.current.getMinutes().toString().padStart(2, '0') +
      '  ' +
      date.current.getDate().toString().padStart(2, '0') +
      '/' +
      monthString.current.toString().padStart(2, '0') +
      '/' +
      date.current.getFullYear().toString()
  );

  async function deletePost() {
    await fetch(`https://${Source.heroku}/deletePost`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        _id: post._id,
      }),
    })
      .then((res) => {
        status.current = res.status;
        return res;
      })
      .then((res) => res.json())
      .then((res) => {
        console.log(res);
        if (status.current === 200) {
          return Alert.alert('Post deleted', 'Your post has been deleted', [
            {
              text: 'OK',
              onPress: () => undefined,
              style: 'cancel',
            },
          ]);
        }
        if (status.current === 422) {
          return Alert.alert(
            'Post not found',
            'The post you want to delete does not exist.',
            [
              {
                text: 'OK',
                onPress: () => undefined,
                style: 'cancel',
              },
            ]
          );
        }
      })
      .catch((err) => console.log(err));
  }

  function askDelete() {
    return Alert.alert(
      'Delete this post?',
      'Are you sure you would like to delete this post? This action is irreversible.',
      [
        {
          text: 'No',
          onPress: () => undefined,
          style: 'cancel',
        },
        {
          text: 'Yes',
          onPress: () => deletePost(),
          style: 'destructive',
        },
      ]
    );
  }

  return (
    <View style={Style.profilePost}>
      <View>
        <Grid>
          <Col style={{ flexDirection: 'row', marginTop: 15 }}>
            <Image
              style={{
                width: 32,
                height: 32,
                marginHorizontal: 8,
                marginTop: 4,
                borderRadius: 28,
              }}
              source={require('../assets/images/profile.png')}
            />
            <View style={{ flexDirection: 'column' }}>
              <Text style={{ fontSize: 18, fontWeight: 'bold' }}>
                {postUsername.current}
              </Text>
              <Text style={{ fontSize: 12 }}>{dateString.current}</Text>
            </View>
          </Col>
          <Col>
            {showButton ? (
              <View
                style={{
                  flexDirection: 'row',
                  alignSelf: 'flex-end',
                  margin: 10,
                }}
              >
                <TouchableOpacity style={{ margin: 8 }} onPress={askDelete}>
                  <FontAwesomeIcon icon={faTrashAlt} />
                </TouchableOpacity>
                <TouchableOpacity
                  style={{ margin: 8 }}
                  onPress={navigation.navigate('EditPost', { post })}
                >
                  <FontAwesomeIcon icon={faEdit} />
                </TouchableOpacity>
                <TouchableOpacity style={{ margin: 8 }} onPress={askDelete}>
                  <FontAwesomeIcon icon={faExclamation} />
                </TouchableOpacity>
              </View>
            ) : (
              <View
                style={{
                  flexDirection: 'row',
                  alignSelf: 'flex-end',
                  margin: 10,
                }}
              >
                <TouchableOpacity style={{ margin: 8 }} onPress={askDelete}>
                  <FontAwesomeIcon icon={faExclamation} />
                </TouchableOpacity>
              </View>
            )}
          </Col>
        </Grid>
      </View>
      <Text
        style={{
          marginHorizontal: 15,
          marginVertical: 10,
          fontSize: 20,
          fontWeight: 'bold',
        }}
      >
        {post.title}
      </Text>
      <Text style={{ marginHorizontal: 15, fontSize: 16 }}>{post.content}</Text>
      <View>
        <LikeAndDislike key={post._id} post={post} />
      </View>
      <View>
        <TouchableOpacity
          style={{ flexDirection: 'row' }}
          onPress={() => setShowComment(!showComment)}
        >
          <Text
            style={{
              alignSelf: 'center',
              margin: 5,
            }}
          >
            Comment
          </Text>
          <FontAwesomeIcon
            icon={faCommentAlt}
            style={{ margin: 5, marginRight: 5 }}
          />
        </TouchableOpacity>
      </View>
      {showComment ? (
        <CommentContainer post={post} state={showComment} />
      ) : null}
    </View>
  );
}
