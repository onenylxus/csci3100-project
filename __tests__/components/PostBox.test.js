// Import
import React from 'react';
import { render } from '@testing-library/react-native';

// Import target component
import PostBox from '../../components/PostBox';

// Mock FontAwesome icons
jest.mock('@fortawesome/react-native-fontawesome', () => ({
  FontAwesomeIcon: '',
}));

// Run jest tests
describe('PostBox', () => {
  let element;

  beforeEach(() => {
    // Mock post
    const postMock = {
      username: 'testac',
      timestamp: new Date(2021, 4, 1),
      numOfLike: 0,
      numOfDislike: 0,
    };

    // Render
    element = render(<PostBox post={postMock} />);
  });

  it('matches snapshot', () => {
    expect(element.toJSON()).toMatchSnapshot();
  });
});
