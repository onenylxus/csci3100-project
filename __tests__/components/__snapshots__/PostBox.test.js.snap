// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostBox matches snapshot 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "backgroundColor": "white",
      "borderColor": "#F0F0F0",
      "borderRadius": 10,
      "borderWidth": 1,
      "flexDirection": "column",
      "justifyContent": "center",
      "marginBottom": 10,
      "marginTop": 20,
      "width": "95%",
    }
  }
>
  <View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginTop": 15,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "borderRadius": 28,
              "height": 32,
              "marginHorizontal": 8,
              "marginTop": 4,
              "width": 32,
            }
          }
        />
        <View
          style={
            Object {
              "flexDirection": "column",
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
            testac
          </Text>
          <Text
            style={
              Object {
                "fontSize": 12,
              }
            }
          >
            0:0  1/4/2021
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "column",
          }
        }
      >
        <View
          style={
            Object {
              "justifyContent": "flex-end",
            }
          }
        >
          <View
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignSelf": "flex-end",
                "margin": 15,
                "opacity": 1,
              }
            }
          >
            <
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f141",
                    "M328 256c0 39.8-32.2 72-72 72s-72-32.2-72-72 32.2-72 72-72 72 32.2 72 72zm104-72c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72zm-352 0c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72z",
                  ],
                  "iconName": "ellipsis-h",
                  "prefix": "fas",
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
  <Text
    style={
      Object {
        "fontSize": 20,
        "fontWeight": "bold",
        "marginHorizontal": 15,
        "marginVertical": 10,
      }
    }
  />
  <Text
    style={
      Object {
        "fontSize": 16,
        "marginHorizontal": 15,
      }
    }
  />
  <View>
    <View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexDirection": "row",
            "opacity": 1,
          }
        }
      >
        <
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f164",
                "M104 224H24c-13.255 0-24 10.745-24 24v240c0 13.255 10.745 24 24 24h80c13.255 0 24-10.745 24-24V248c0-13.255-10.745-24-24-24zM64 472c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zM384 81.452c0 42.416-25.97 66.208-33.277 94.548h101.723c33.397 0 59.397 27.746 59.553 58.098.084 17.938-7.546 37.249-19.439 49.197l-.11.11c9.836 23.337 8.237 56.037-9.308 79.469 8.681 25.895-.069 57.704-16.382 74.757 4.298 17.598 2.244 32.575-6.148 44.632C440.202 511.587 389.616 512 346.839 512l-2.845-.001c-48.287-.017-87.806-17.598-119.56-31.725-15.957-7.099-36.821-15.887-52.651-16.178-6.54-.12-11.783-5.457-11.783-11.998v-213.77c0-3.2 1.282-6.271 3.558-8.521 39.614-39.144 56.648-80.587 89.117-113.111 14.804-14.832 20.188-37.236 25.393-58.902C282.515 39.293 291.817 0 312 0c24 0 72 8 72 81.452z",
              ],
              "iconName": "thumbs-up",
              "prefix": "fas",
            }
          }
          style={
            Object {
              "color": "lightgrey",
              "margin": 5,
            }
          }
        />
        <Text
          style={
            Object {
              "alignSelf": "center",
              "marginRight": 15,
            }
          }
        >
          0
        </Text>
      </View>
    </View>
  </View>
  <View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "margin": 5,
          }
        }
      >
        Comment
      </Text>
      <
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f27a",
              "M448 0H64C28.7 0 0 28.7 0 64v288c0 35.3 28.7 64 64 64h96v84c0 9.8 11.2 15.5 19.1 9.7L304 416h144c35.3 0 64-28.7 64-64V64c0-35.3-28.7-64-64-64z",
            ],
            "iconName": "comment-alt",
            "prefix": "fas",
          }
        }
        style={
          Object {
            "margin": 5,
            "marginRight": 5,
          }
        }
      />
    </View>
  </View>
</View>
`;
